onvif/src/discovery/mod.rs

```diff
@@ -81,52 +87,49 @@ pub async fn discover(duration: std::time::Duration) -> Result<impl Stream<Item
         let local_socket_addr = SocketAddr::new(IpAddr::V4(LOCAL_IPV4_ADDR), LOCAL_PORT);
         let multi_socket_addr = SocketAddr::new(IpAddr::V4(MULTI_IPV4_ADDR), MULTI_PORT);
 
-        let socket = async_std::net::UdpSocket::bind(local_socket_addr).await?;
+        let socket = UdpSocket::bind(local_socket_addr).await?;
         socket.join_multicast_v4(MULTI_IPV4_ADDR, LOCAL_IPV4_ADDR)?;
         socket
             .send_to(&probe_xml.as_bytes(), multi_socket_addr)
             .await?;
 
-        Ok::<async_std::net::UdpSocket, std::io::Error>(socket)
-    })()
-    .await?;
+        socket
+    };
 
-    let stream = {
-        use futures_util::stream::StreamExt;
+    Ok(stream! {
+        let probe = &probe;
+        let socket = &socket;
 
-        // Make an async stream of XML's
-        futures_util::stream::unfold(socket, |s| async { Some((recv_string(&s).await, s)) })
-            .filter_map(|string| async move { string.ok() })
-            .filter_map(|xml| async move {
-                debug!("Probe match XML: {}", xml);
-                yaserde::de::from_str::<probe_matches::Envelope>(&xml).ok()
-            })
-            .filter(move |envelope| {
-                futures_util::future::ready(envelope.header.relates_to == probe.header.message_id)
-            })
-            .filter_map(|envelope| get_responding_addr(envelope, is_addr_responding))
-    };
+        let start = Instant::now();
+        loop {
+            let elapsed = start.elapsed();
+            if start.elapsed() >= duration {
+                break;
+            }
 
-    let timed_stream = {
-        use async_std::stream::StreamExt;
-
-        stream
-            // Blend in an interval stream to implement timeout
-            // Let our payload stream contain Some's and timeout stream contain None's
-            .map(Some)
-            .merge(async_std::stream::interval(duration).map(|_| None))
-            // Terminate stream when the first None is received
-            .take_while(|event| event.is_some())
-            .filter_map(|event| event)
-    };
+            let timeout = duration - elapsed;
 
-    Ok(timed_stream)
+            // Separate async block to be able to short-circuit on `None`.
+            let try_produce_item = async move {
+                let xml = recv_string(socket, timeout).await.ok()?;
+                debug!("Probe match XML: {}", xml);
+                let envelope = yaserde::de::from_str::<probe_matches::Envelope>(&xml).ok()?;
+                if envelope.header.relates_to != probe.header.message_id {
+                    return None;
+                }
+                get_responding_addr(envelope, is_addr_responding).await
+            };
+
+            if let Some(item) = try_produce_item.await {
+                yield item;
```

Comment:
Is `yield` already on stable?

Reply:
No, the `stream!` macro transforms it into something else (probably a function call + `.await`).